src/System.CommandLine/Parsing/ParseResultVisitor.cs

```diff
@@ -179,31 +162,33 @@ private void VisitOptionNode(OptionNode optionNode)
                     optionNode.Token,
                     _innermostCommandResult);
 
-                AddToResult(optionResult);
+                _symbolResults.Add(optionNode.Option, optionResult);
+
+                if (optionNode.Children is null) // no Arguments
+                {
+                    if (optionResult.Option.Argument.HasCustomParser)
```

Comment:
previously this logic was performed as part of the validation (`ValidateAndConvertOptionResult`). Since here we know that given `OptionNode` was parsed with no arguments, we can create the argument result now and let validation method just validate without performing mutations.

https://github.com/dotnet/command-line-api/blob/9c48247db71bfbf7309ec48d87929a8fe425a64d/src/System.CommandLine/Parsing/ParseResultVisitor.cs#L489-L500